File: Task Tracker (GO)/storage.go
package main

import (
	"encoding/json"
	"os"
)

type Storage interface {
	Load() ([]Task, error)
	Save(tasks []Task) error
}

type JSONStorage struct {
	filename string
}

func NewJSONStorage(filename string) *JSONStorage {
	return &JSONStorage{filename: filename}
}

func (s *JSONStorage) Load() ([]Task, error) {
	data, err := os.ReadFile(s.filename)
	if err != nil {
		if os.IsNotExist(err) {
			return []Task{}, nil
		}
		return nil, err
	}

	var tasks []Task
	if err := json.Unmarshal(data, &tasks); err != nil {
		return nil, err
	}
	return tasks, nil
}

func (s *JSONStorage) Save(tasks []Task) error {
	data, err := json.MarshalIndent(tasks, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(s.filename, data, 0644)
}

File: TaskTrackerRust/task_tracker_rust/src/storage.rs
use crate::task::Task;
use serde_json;
use std::fs;
use std::io;
use std::path::Path;

pub trait Storage {
    fn load(&self) -> io::Result<Vec<Task>>;
    fn save(&self, tasks: &Vec<Task>) -> io::Result<()>;
}

pub struct JSONStorage {
    pub filename: String,
}

impl JSONStorage {
    pub fn new(filename: &str) -> Self {
        JSONStorage {
            filename: filename.to_string(),
        }
    }
}

impl Storage for JSONStorage {
    fn load(&self) -> io::Result<Vec<Task>> {
        if !Path::new(&self.filename).exists() {
            return Ok(vec![]);
        }
        let data = fs::read_to_string(&self.filename)?;
        let tasks: Vec<Task> = serde_json::from_str(&data)?;
        Ok(tasks)
    }

    fn save(&self, tasks: &Vec<Task>) -> io::Result<()> {
        let data = serde_json::to_string_pretty(tasks)?;
        fs::write(&self.filename, data)?;
        Ok(())
    }
}
